src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Serialization/JsonDataReader.cs

```diff
@@ -0,0 +1,528 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.ProjectEngineHost.Serialization;
+
+internal delegate void ReadPropertyValue<TData>(JsonDataReader reader, ref TData data);
+internal delegate T ReadValue<T>(JsonDataReader reader);
+internal delegate T ReadProperties<T>(JsonDataReader reader);
+internal delegate void ProcessValue<T>(JsonDataReader reader, T arg);
+internal delegate void ProcessProperties<T>(JsonDataReader reader, T arg);
+
+internal sealed class PropertyMap<TData>
+    where TData : struct
+{
+    private readonly Dictionary<string, ReadPropertyValue<TData>> _map;
+
+    public PropertyMap(params (string, ReadPropertyValue<TData>)[] pairs)
+    {
+        var map = new Dictionary<string, ReadPropertyValue<TData>>(capacity: pairs.Length);
+
+        foreach (var (key, value) in pairs)
+        {
+            map.Add(key, value);
+        }
+
+        _map = map;
+    }
+
+    public bool TryGetPropertyReader(string key, [MaybeNullWhen(false)] out ReadPropertyValue<TData> value)
+        => _map.TryGetValue(key, out value);
+}
+
+internal partial class JsonDataReader
```

Comment:
What's the relationship between JsonDataReader and JsonReader? This could use some documentation (I'm not done with the PR so may be more obvious reading code...)

Reply:
JsonDataReader is an abstraction that wraps a JSON.NET JsonReader. I'll add a doc.